packages/agoric-cli/src/install.js

```diff
@@ -40,7 +40,117 @@ export default async function installMain(progname, rawArgs, powers, opts) {
   }
 
   let subdirs;
+  const installDirs = ['.'];
   let sdkWorktree;
+  const versions = new Map();
+  const linkFolder = path.resolve(`_agstate/yarn-links`);
+  const linkFlags = [];
+  if (opts.sdk) {
+    linkFlags.push(`--link-folder=${linkFolder}`);
+  }
+
+  const doYarnInstalls = async (phase, ...flags) => {
+    for (const instdir of installDirs) {
+      log.info(`yarn install ${phase} in ${instdir}`);
+      // eslint-disable-next-line no-await-in-loop
+      const yarnInstall = await pspawn(
+        'yarn',
+        [...linkFlags, 'install', ...flags],
+        {
+          stdio: 'inherit',
+          cwd: instdir,
+        },
+      );
+      if (yarnInstall) {
+        throw Error(`yarn install ${phase} failed in ${instdir}`);
+      }
+    }
+  };
+
+  const updateSdkVersion = async forceVersion => {
+    // Prune the old version, and update to the new version.
+    const updatesTodo = [];
+    const pruningTodo = [];
+
+    await Promise.all(
+      subdirs.map(async subdir => {
+        // Modify all the SDK package versions.
+        const pjson = `${subdir}/package.json`;
+        const packageJSON = await fs
+          .readFile(pjson, 'utf-8')
+          .catch(_ => undefined);
+        if (!packageJSON) {
+          return;
+        }
+        const pj = JSON.parse(packageJSON);
+        const prunedPj = { ...pj };
+        let needsPruning = false;
+        for (const section of ['dependencies', 'devDependencies']) {
+          const deps = pj[section];
+          if (deps) {
+            const prunedDeps = { ...deps };
+            prunedPj[section] = prunedDeps;
+            for (const pkg of Object.keys(deps)) {
+              if (versions.has(pkg)) {
+                if (deps[pkg] === forceVersion) {
+                  // We need to remove the old package and install to prune the
+                  // old dependencies.  Only after that can we add the new deps
+                  // and run install to pick them up.
+                  delete prunedDeps[pkg];
+                  needsPruning = true;
+                }
+                deps[pkg] = forceVersion;
+              }
+            }
+          }
+        }
+        // Ensure we update the package.json before exiting.
+        const updatePackageJson = async () => {
+          // Don't update on exit anymore.
+          process.off('beforeExit', updatePackageJson);
+          log.info(`updating ${pjson}`);
+          await fs.writeFile(pjson, `${JSON.stringify(pj, null, 2)}\n`);
+        };
+        updatesTodo.push(updatePackageJson);
+        if (needsPruning) {
+          pruningTodo.push(async () => {
+            // Update on exit, in case we are interrupted.
+            process.on('beforeExit', updatePackageJson);
```

Comment:
> I presume you considered and decided against framing this in terms of a finally. I don’t think it would occur to me to use the beforeExit event under any circumstances.

It *is* framed in terms of a `pruningP.finally(...)`.  It's quite important that these updates are done even if the process is interrupted.  Do you have a better mechanism at hand?

Reply:
No, there’s nothing better in the face of uncaught exceptions.